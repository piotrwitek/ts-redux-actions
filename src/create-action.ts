import { PayloadAction } from '.';

export type ActionCreatorFunction<TS extends string> =
  | ((...args: any[]) => { type: TS, payload?: any, meta?: any, error?: boolean });

export type TSActionCreator<AC, T extends string> = AC & {
  readonly type: T,
};

export function createAction<TS extends string, AC extends (() => { type: TS })>(
  typeString: TS,
): TSActionCreator<AC, TS>;

export function createAction<TS extends string, AC extends ActionCreatorFunction<TS>>(
  typeString: TS,
  creatorFunction: AC,
): TSActionCreator<AC, TS>;

export function createAction<TS extends string, AC extends ActionCreatorFunction<TS>>(
  typeString: TS,
  creatorFunction?: AC,
): TSActionCreator<AC, TS> {
  if (creatorFunction) {
    if (typeof creatorFunction !== 'function') {
      throw new Error('second argument is not a function');
    }

    const actionCreator: any = creatorFunction;
    actionCreator.type = typeString;

    return actionCreator;
  } else {
    const actionCreator: any =
      () => ({ type: typeString });
    actionCreator.type = typeString;

    return actionCreator;
  }
}

export function createPayloadAction<TS extends string, T extends {}>(typeString: TS, props: T) {
    function creator(val: T): PayloadAction<TS, T> {
       return {
         type: typeString,
         payload: {...props as object, ...val as object} as T,
       };
    }
    return createAction(typeString, creator);
  }
