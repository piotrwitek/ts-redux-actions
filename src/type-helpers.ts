/**
 * PUBLIC API
 */

/**
 * @desc Infers Action union-type from action-creator map object
 */
export type ActionType<
  ActionCreatorOrMap
> = ActionCreatorOrMap extends ActionCreator<StringType>
  ? ReturnType<ActionCreatorOrMap>
  : ActionCreatorOrMap extends object
  ? ActionCreatorMap<ActionCreatorOrMap>[keyof ActionCreatorOrMap]
  : never;

/**
 * @desc Infers State object from reducer map object
 */
export type StateType<ReducerOrMap> = ReducerOrMap extends (
  ...args: any[]
) => any
  ? ReturnType<ReducerOrMap>
  : ReducerOrMap extends object
  ? { [K in keyof ReducerOrMap]: StateType<ReducerOrMap[K]> }
  : never;

/**
 * INTERNAL API
 */

/**
 * @private
 * @desc Representing action-type of string
 */
export type StringType = string;

/**
 * @private
 * @desc Action without Payload
 * @type T - ActionType
 */
export type EmptyAction<T extends StringType> = {
  type: T;
};

/**
 * @private
 * @desc Action with only Payload
 * @type T - ActionType
 * @type P - Payload
 */
export type PayloadAction<T extends StringType, P> = {
  type: T;
  payload: P;
};

/**
 * @private
 * @desc Action with only Meta
 * @type T - ActionType
 * @type P - Payload
 * @type M - Meta
 */
export type MetaAction<T extends StringType, M> = {
  type: T;
  meta: M;
};

/**
 * @private
 * @desc Action with both Payload and Meta
 * @type T - ActionType
 * @type P - Payload
 * @type M - Meta
 */
export type PayloadMetaAction<T extends StringType, P, M> = {
  type: T;
  payload: P;
  meta: M;
};

/**
 * TODO: NOT USED
 * @private
 * @desc Flux Standard Action
 * @type T - ActionType
 * @type P - Payload
 * @type M - Meta
 */
export interface FluxStandardAction<T extends StringType, P = void, M = void> {
  type: T;
  payload: P;
  meta: M;
  error?: true;
}

/** @private */
export interface TypeMeta<T extends StringType> {
  getType?: () => T;
}

/** @private */
export type Box<T> = { v: T };

/** @private */
export type Unbox<T extends Box<any>> = T['v'];

/** @private */
export type NoArgCreator<T extends StringType> = () => EmptyAction<T>;

/** @private */
export type PayloadCreator<T extends StringType, P> = (
  payload: P
) => PayloadAction<T, P>;

/** @private */
export type PayloadMetaCreator<T extends StringType, P, M> = (
  payload: P,
  meta: M
) => PayloadMetaAction<T, P, M>;

/** @private */
export type FsaBuilder<
  T extends StringType,
  P extends Box<any> = Box<void>,
  M extends Box<any> = Box<void>
> = M extends Box<void>
  ? P extends Box<void>
    ? NoArgCreator<T>
    : PayloadCreator<T, Unbox<P>>
  : PayloadMetaCreator<T, Unbox<P>, Unbox<M>>;

/** @private */
export type FsaMapBuilder<
  T extends StringType,
  Result extends Box<any>,
  PArg extends Box<any> = Box<void>,
  MArg extends Box<any> = Box<void>
> = MArg extends Box<void>
  ? PArg extends Box<void>
    ? () => { type: T } & Unbox<Result>
    : (payload: Unbox<PArg>) => { type: T } & Unbox<Result>
  : (payload: Unbox<PArg>, meta: Unbox<MArg>) => { type: T } & Unbox<Result>;

/** @private */
export type ActionCreator<T extends StringType> = (
  ...args: any[]
) => { type: T };

/** @private */
export type ActionCreatorMap<T> = { [K in keyof T]: ActionType<T[K]> };
