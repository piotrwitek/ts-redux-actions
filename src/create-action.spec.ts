import * as T from './type-helpers';
import { createAction } from './create-action';

describe('toString() method return a type', () => {
  const actionCreator = createAction('CREATE_ACTION');
  // @dts-jest:pass:snap
  actionCreator.toString(); // => 'CREATE_ACTION'
});

describe('with symbol', () => {
  const CREATE_ACTION = Symbol(1);
  const withSymbol = createAction(CREATE_ACTION as any);
  // @dts-jest:pass:snap
  withSymbol(); // => { type: CREATE_ACTION }
});

describe('with type only', () => {
  const withTypeOnly = createAction('CREATE_ACTION');
  // @dts-jest:pass:snap
  withTypeOnly(); // => { "type": "CREATE_ACTION" }
});

describe('with payload', () => {
  const withPayload = createAction('CREATE_ACTION', resolve => {
    return (id: number) => resolve(id);
  });
  // @dts-jest:pass:snap
  withPayload(1); // => { type: 'CREATE_ACTION', payload: 1 }
});

// TODO: optionals could be solved with overloads
describe('with optional payload', () => {
  const withPayload = createAction('CREATE_ACTION', resolve => {
    return (id?: number) => resolve(id);
  });
  // @dts-jest:pass:snap
  withPayload(); // => { type: 'CREATE_ACTION' }
  // @dts-jest:pass:snap
  withPayload(1); // => { type: 'CREATE_ACTION', payload: 1 }
});

describe('with meta', () => {
  const withMeta = createAction('CREATE_ACTION', resolve => {
    return (token: string) => resolve(undefined, token);
  });
  // @dts-jest:pass:snap
  withMeta('token'); // => { type: 'CREATE_ACTION', meta: 'token' }
});

describe('with payload and meta', () => {
  const withPayloadAndMeta = createAction('CREATE_ACTION', resolve => {
    return (id: number, token: string) => resolve(id, token);
  });
  // @dts-jest:pass:snap
  withPayloadAndMeta(1, 'token'); // => { type: 'CREATE_ACTION', payload: 1, meta: 'token' }
});

describe('type-safe usage with higher-order function', () => {
  interface UserSettingsState {
    settingA: string;
    settingB: number;
  }

  const setUserSetting = <K extends keyof UserSettingsState>(
    setting: K,
    newValue: UserSettingsState[K]
  ) =>
    createAction('SET_USER_SETTING', resolve => () =>
      resolve({ setting, newValue })
    )();

  // @dts-jest:pass:snap
  setUserSetting('settingA', 'foo');
  // @dts-jest:fail:snap
  setUserSetting('settingA', 0); // Error as expected

  // @dts-jest:pass:snap
  setUserSetting('settingB', 0);
  // @dts-jest:fail:snap
  setUserSetting('settingB', 'foo'); // Error as expected
});
