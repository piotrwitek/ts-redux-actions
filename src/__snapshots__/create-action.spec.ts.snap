// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actionCreator.toString() 1`] = `"string"`;

exports[`setUserSetting('settingA', 'foo') 1`] = `"{ type: \\"SET_USER_SETTING\\"; payload: { setting: \\"settingA\\"; newValue: string; }; }"`;

exports[`setUserSetting('settingA', 0) 1`] = `"Argument of type '0' is not assignable to parameter of type 'string'."`;

exports[`setUserSetting('settingB', 'foo') 1`] = `"Argument of type '\\"foo\\"' is not assignable to parameter of type 'number'."`;

exports[`setUserSetting('settingB', 0) 1`] = `"{ type: \\"SET_USER_SETTING\\"; payload: { setting: \\"settingB\\"; newValue: number; }; }"`;

exports[`withMeta('token') 1`] = `"{ type: \\"CREATE_ACTION\\"; meta: string; }"`;

exports[`withPayload() 1`] = `"{ type: \\"CREATE_ACTION\\"; payload: number; }"`;

exports[`withPayload(1) 1`] = `"{ type: \\"CREATE_ACTION\\"; payload: number; }"`;

exports[`withPayload(1) 2`] = `"{ type: \\"CREATE_ACTION\\"; payload: number; }"`;

exports[`withPayloadAndMeta(1, 'token') 1`] = `"{ type: \\"CREATE_ACTION\\"; payload: number; meta: string; }"`;

exports[`withSymbol() 1`] = `"{ type: any; }"`;

exports[`withTypeOnly() 1`] = `"{ type: \\"CREATE_ACTION\\"; }"`;
